hbase-server/src/main/java/org/apache/hadoop/hbase/mapreduce/LoadIncrementalHFilesJob.java

```diff
@@ -0,0 +1,348 @@
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hbase.HBaseConfiguration;
+import org.apache.hadoop.hbase.HColumnDescriptor;
+import org.apache.hadoop.hbase.TableName;
+import org.apache.hadoop.hbase.client.Connection;
+import org.apache.hadoop.hbase.client.ConnectionFactory;
+import org.apache.hadoop.hbase.client.RegionLocator;
+import org.apache.hadoop.hbase.client.Table;
+import org.apache.hadoop.hbase.io.HFileLink;
+import org.apache.hadoop.hbase.regionserver.StoreFileInfo;
+import org.apache.hadoop.hbase.util.Bytes;
+import org.apache.hadoop.io.ArrayWritable;
+import org.apache.hadoop.io.NullWritable;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.io.Writable;
+import org.apache.hadoop.mapreduce.*;
+import org.apache.hadoop.mapreduce.lib.output.NullOutputFormat;
+import org.apache.hadoop.util.Tool;
+import org.apache.hadoop.util.ToolRunner;
+
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.IOException;
+import java.util.*;
+
+public class LoadIncrementalHFilesJob extends Configured implements Tool {
+
+  private static final Log LOG = LogFactory.getLog(LoadIncrementalHFilesJob.class);
+
+  public static final String ROOT_DIR = "loadincrementalhfilesjob.root.dir";
+  public static final String TABLE_NAME = "loadincrementalhfilesjob.table.name";
+  public static final String MAX_MAP_TASK = "loadincrementalhfilesjob.max.map.tasks";
+  public static final String DEPTH = "loadincrementalhfilesjob.depth";
+
+  @Override public int run(String[] args) throws Exception {
+    Configuration configuration = HBaseConfiguration.create(getConf());
+
+    return run(args[0], args[1], configuration);
+  }
+
+  @VisibleForTesting public int run(String rootPath, String table, Configuration configuration)
+      throws Exception {
+    Path basePath = new Path(rootPath);
+    configuration.set(TABLE_NAME, table);
+    configuration.set(ROOT_DIR, basePath.toString());
+
+    if (!bulkload(basePath, table, configuration)) {
+      return 1;
+    }
+
+    return 0;
+  }
+
+  private boolean bulkload(Path rootDir, String table, Configuration configuration)
+      throws Exception {
+    Configuration jobConfiguration = new Configuration(configuration);
+
+    applyNonOverridableOptions(jobConfiguration);
+
+    Job job = Job.getInstance(jobConfiguration, "Bulkload-" + table + "-" + rootDir.toString());
+
+    TableMapReduceUtil.addDependencyJars(job);
+
+    job.setJarByClass(LoadIncrementalHFilesJob.class);
+
+    job.setInputFormatClass(BulkLoadInputFormat.class);
+    job.setOutputFormatClass(NullOutputFormat.class);
+
+    job.setMapperClass(BulkoadMapper.class);
+    job.setMapOutputKeyClass(NullWritable.class);
+    job.setMapOutputValueClass(NullWritable.class);
+
+    job.setNumReduceTasks(0);
+
+    return job.waitForCompletion(true);
+  }
+
+  private static void applyNonOverridableOptions(Configuration configuration) {
```

Comment:
These settings basically account for the fact that the map task is not using a hadoop `OutputFormat` to write its output: 
 - Task timeout is disabled because it is based on the map/reduce idea of progress (which means either writing to the hadoop `Mapper.Context` or the `OutputFormat`). The `BulkloadMapper` code does neither, so it would inevitably timeout.
 - Speculative execution and task retry are disabled because they would fail anyway. A speculated task would attempt to bulkload the same set of files as the original resulting in a failure. A retried task would potentially attempt to bulkload files that were already bulk loaded by the original task.

I don't think an `OutputFormat` can be implemented because there is no way to implement the `OutputCommitter` contract reliably without changes to the region server side.

Reply:
> Task timeout is disabled because it is based on the map/reduce idea of progress (which means either writing to the hadoop Mapper.Context or the OutputFormat). The BulkloadMapper code does neither, so it would inevitably timeout.

What if you make BulkloadMapper update mapper context status after calling `loadIncrementalHFiles.doBulkloadFromQueue`? I don't feel good about completely disabling timeouts, and making it non configurable, this could lead to yarn pool resources being hijacked permanently until an operator manually intervention to kill the slow bulkload job.

> Speculative execution and task retry are disabled because they would fail anyway. A speculated task would attempt to bulkload the same set of files as the original resulting in a failure. A retried task would potentially attempt to bulkload files that were already bulk loaded by the original task

Agreed about speculative execution, but not for retries. I would rather re-bulkload a few hfiles for a given failed task, then having to redo the whole bulkload from the scratch.

And how about `mapreduce.task.classpath.user.precedence`? Is it to make sure this version of LoadIncrementalHFiles gets into the task classpath, when running on clusters that didn't have this patch? Am fine with it, if that's the case, just wanted to confirm.

Reply:
> What if you make BulkloadMapper update mapper context status after calling loadIncrementalHFiles.doBulkloadFromQueue? I don't feel good about completely disabling timeouts, and making it non configurable, this could lead to yarn pool resources being hijacked permanently until an operator manually intervention to kill the slow bulkload job.

That's a good point. The `LoadIncrementalHFiles` class should, however, be changed to allow loading a file one by one and re-use the underlying HBase connection. If we just update the mapper context after the call to `loadIncrementalHFiles.doBulkloadFromQueue ` it still won't have any effect because that is the last call in the mapper's lifetime. I'll try and incorporate this change.


> Agreed about speculative execution, but not for retries. I would rather re-bulkload a few hfiles for a given failed task, then having to redo the whole bulkload from the scratch.

You can't actually re-bulkload a file because the API is not idempotent (the region server moves the file into the region directory, a subsequent call with the same file will throw a file not found error). However, we can add a file existence check on the mapper side so a map task can be re-tried. Does that sound good?

> And how about mapreduce.task.classpath.user.precedence? Is it to make sure this version of LoadIncrementalHFiles gets into the task classpath, when running on clusters that didn't have this patch? Am fine with it, if that's the case, just wanted to confirm.

That's just a setting particular to a setup I used for testing. I'll remove it and allow the client to specify it, if needed.

Reply:
> The LoadIncrementalHFiles class should, however, be changed to allow loading a file one by one and re-use the underlying HBase connection.

Or maybe your mapper could use `bulkLoadPhase` directly, in order to have more control of which lqis got successfully loaded, and which didn't.

> You can't actually re-bulkload a file because the API is not idempotent (the region server moves the file into the region directory, a subsequent call with the same file will throw a file not found error). 

IIRC, it would move to a temp dir first, if the load fails, file will be left there in the pending dir, no? In that case, operators could react upon what is left there in the temp dir.

> However, we can add a file existence check on the mapper side so a map task can be re-tried. Does that sound good?

Task retry logic is a good addition, but each task should still have timeouts and a maximum number of attempts limit. What I meant in my previous comment is that it should not hang forever, as this might impact other yarn jobs on same pool.

Reply:
> IIRC, it would move to a temp dir first, if the load fails, file will be left there in the pending dir, no? In that case, operators could react upon what is left there in the temp dir.

Afaik, the temp directory is only used for the secure bulkload feature.

Reply:
> Task retry logic is a good addition, but each task should still have timeouts and a maximum number of attempts limit. What I meant in my previous comment is that it should not hang forever, as this might impact other yarn jobs on same pool.

I have refactored things a bit. The call to `doBulkloadFromQueue` is moved in the map method and we now use a listener for the bulk loaded files so we can notify the Hadoop framework of progress.
 The infinite task timeout was dropped.

I still kept the one attempt per map limit. This can be removed as well and replaced with a file existence check on the mapper side, but it seems hard to distinguish between two situations in this case: the file does not exist because this is another attempt of a map and that file was already loaded _and_ there is a bug somewhere else that cause the file not to be found.